lightningd/payalgo.c

```diff
@@ -15,6 +16,107 @@
 #include <lightningd/subd.h>
 #include <sodium/randombytes.h>
 
+/* Record of failures. */
+enum pay_failure_type {
+	FAIL_UNPARSEABLE_ONION,
+	FAIL_PAYMENT_REPLY
+};
+/* A payment failure. */
+struct pay_failure {
+	/* Part of a list of failures */
+	struct list_node list;
+	/* The type of payment failure */
+	enum pay_failure_type type;
+	/* A tal_arr of route hops, whose parent is
+	 * this struct */
+	struct route_hop *route;
+	/* The raw onion reply, if TYPE_UNPARSEABLE_ONION, a
+	 * tal_arr whose parent is this struct */
+	const u8 *onionreply;
+	/* The routing failure, if TYPE_PAYMENT_REPLY, a tal
+	 * object whose parent is this struct */
+	struct routing_failure *routing_failure;
+};
+
+/* FIXME: move json_add_route_hop and json_add_route to
+ * common/json.c, share code with getroute */
```

Comment:
Use lightningd/jsonrpc.h where we put the other lightningd-only JSON helpers.  Or, better, refactor those into lightningd/json.[ch].